caffe2/operators/recurrent_network_op_gpu.cu

```diff
@@ -0,0 +1,148 @@
+#include "caffe2/core/context_gpu.h"
+#include "caffe2/operators/recurrent_network_op.h"
+
+namespace caffe2 {
+
+namespace detail {
+
+template <typename T, typename Context>
+void initializeRecurrentInput(
+    const RecurrentInput& rc,
+    int32_t seqLen,
+    int32_t batchSize,
+    Workspace* ws,
+    Context* context);
+
+namespace {
+
+template <typename T>
+__global__
+void initRecurrentInput_kernel(
+    size_t stateSize,
+    const T* input,
+    T* state) {
+  // index into appropriate target buffer
+  const int block_id = blockIdx.x;
+  T* state_local = state + block_id*stateSize;
+
+  // copy
+  for (int idx=threadIdx.x; idx < stateSize; idx+=blockDim.x) {
+    state_local[idx] = input[idx];
+  }
+}
+
+
+}; // namespace
+
+template <>
+void initializeRecurrentInput<float,CUDAContext>(
+    const RecurrentInput& rc,
+    int32_t seqLen,
+    int32_t batchSize,
+    Workspace* ws,
+    CUDAContext* context) {
+  auto stateBlob = ws->GetBlob(rc.state);
+  CAFFE_ENFORCE(stateBlob);
+  auto* state = stateBlob->GetMutable<Tensor<CUDAContext>>();
+
+  auto inputBlob = ws->GetBlob(rc.input);
+  CAFFE_ENFORCE(inputBlob);
+  const auto& input = inputBlob->Get<Tensor<CUDAContext>>();
+  CAFFE_ENFORCE_GE(input.ndim(), 1, rc.input);
+  CAFFE_ENFORCE_LE(input.ndim(), 3, rc.input);
+
+  const auto stateSize = input.dim(input.ndim() - 1);
+  // States at [0, ..., T] (inclusive)
+  state->Resize(seqLen + 1, batchSize, stateSize);
+
+  if (input.ndim() == 3) {
+    CAFFE_ENFORCE_EQ(input.dim(0), 1, rc.input);
+  }
+  if (input.ndim() >= 2) {
+    CAFFE_ENFORCE_EQ(input.dim(input.ndim() - 2), batchSize, rc.input);
+    context->Copy<float, CUDAContext, CUDAContext>(
+        batchSize * stateSize,
+        input.data<float>(),
+        state->mutable_data<float>());
+  } else {
+    initRecurrentInput_kernel<float><<<batchSize, CAFFE_CUDA_NUM_THREADS, 0, context->cuda_stream()>>>(
+        stateSize,
+        input.data<float>(),
+        state->mutable_data<float>());
+  }
+}
+template <>
+void initializeRecurrentInput<float16,CUDAContext>(
```

Comment:
It's just because we have to specialize the function for CUDAContext, which then requires to specify a datatype as well. I wish we had partial specialization for functions!

Reply:
I see the difference is only in the one line (i.e. using a kernel instead of doing Copy in a loop). Could you instead of replicating this function 3 times, implement 3 times a smaller function, which does only replication logic? Also you could make it a template <class Context> class with a static template <typename T> replicate method. Then you could implement this function only two times, not 3 times.

I.e. some examples here: https://ideone.com/p3VgHA

Also was it really a bottleneck? (calling Copy in a loop). In our previous results it was negligible, I think.

Reply:
We had a case where this loop was taking too much time and it was improved by ~300x using the kernel. This was about 6 months ago and I don't remember the test but I can try to reproduce it if it's necessary?

Reply:
@salexspb  I agree with removing the duplication. Let me try that.

Reply:
@salexspb Could you review how I removed the duplication. Let me know if it looks good, then I will `git squash` the last commit.